### ISSUE TITLE
AsyncLocal lost in Test

### ISSUE THREAD

**Issue by Mauzer93** [xunit/xunit#1880]
I am using AsyncLocal<T> to have an isolated context between different tests.
1 out of 10 times, the AsyncLocal that's configured for the current test, seems to lose it's context within the same test.

Why is this happening? How can this be solved? Normally, the AsyncLocal should be the same within the same execution context, so basically within every test?

**Comment by bradwilson** [comment]
Can you provide an example test that reproduces the problem?

**Comment by Mauzer93** [comment]
I have a class container which registers and resolve instances from
`private static AsyncLocal<ConcurrentDictionary<Type, object>> _asyncLocalDictionary = new AsyncLocal<ConcurrentDictionary<Type, object>>();`

DomainStore.Raise(Event), this method does the following:

- Find handler corresponding to Event
- Ask container above to resolve the Handler
- Events.Add(Task.Run(() => {Handler.Handle(event)}).ContinueWith(t => Events.TryTake(t));

In my unit test I have code simular to:
// 
- Container.Register(HandlerInstance)
- DomainStore .Handle(new Event())
- Task.WhenAll(DomainStore.Events).Wait();

=> Registered Handler instance is found most of times, but not always.

**Comment by bartelink** [comment]
@Mauzer93 tests in a single class never run concurrently - looks like you have the answer I was going to give from Stephen's response - yay github!

**Comment by Mauzer93** [comment]
@bartelink in that specific test case, there is only one test method in that class that uses this logic.
The strange thing is that the logic is used in other test classes, where they do have multiple test methods using this async container, and those test have never failed (yet)...

**Comment by bradwilson** [comment]
There is no guarantee that any two tests will share the same async context, even if they're in the same test class. The only guarantee is that a single test method will have a consistent async context (unless the developer does something to break that guarantee, like using `.ConfigureAwait(false)`.

If you want a guarantee of being able to share something across tests in the same class, you must use a [class fixture](https://xunit.github.io/docs/shared-context#class-fixture).

**Comment by Mauzer93** [comment]
I have created a repro project.
In my own project, I see that in 1/x tries, the handler was not found in my AsyncLocal container.

Because I am using AsyncLocal and Unit test classes should have their own Async Context, I assert against the amount of registrations to make the tests fail more easy.
https://github.com/Mauzer93/AsyncLocalXUnit/tree/master/ReproduceAsyncLocalIssueXUnit

**Comment by bradwilson** [comment]
Your use of `Task.Run()` causes the handlers to be called on random thread pool threads, each with different execution contexts. `Task.Run()` does NOT preserve execution context, and as such, you cannot expect `AsyncLocal<T>` to be useful inside any code that's in a `Task.Run`.

**Comment by Mauzer93** [comment]
Task.Run() runs on a random thread, but the handler is resolved before the Task.Run, so the handler should be found correctly as far as I understand it. The method on the handler should not necessarly be executed in the same execution context.

**Comment by Mauzer93** [comment]
According to this article https://riptutorial.com/dot-net/example/14971/flowing-execution-context-with-asynclocal the Execution Context should flow in Task.Run?

**Comment by bradwilson** [comment]
You are overwriting a global static value (`DomainEventStore.InstanceFactory`) in every test, even when those tests are running in parallel with each other, so the result of retrieving the instance factory from any given test is not guaranteed to be the same one that that test set.

**Comment by Mauzer93** [comment]
That's correct. However, the global static InstanceFactory, points to the same AsyncLocal static dictionary. Which should always resolve to right handlers?

**Comment by Mauzer93** [comment]
Thanks for you replies. The sample was not representative for my issue. We worked around by using another implemenation in the Domain Event Store.

**Comment by tmenier** [comment]
I'm not sure if this is the same issue or a new one, but I've discovered that `AsyncLocal` doesn't appear to "hold" when a value is set in `InitializeAsync`. Here is very simple repro:

```c#
public class AsyncLocalTests : IAsyncLifetime
{
    static AsyncLocal<string> _asyncLocal1 = new AsyncLocal<string>();
    static AsyncLocal<string> _asyncLocal2 = new AsyncLocal<string>();

    public AsyncLocalTests() {
        _asyncLocal1.Value = "foo";
    }

    public async Task InitializeAsync() {
        _asyncLocal2.Value = "foo";
    }

    [Fact]
    public void WorksWithSyncSetup() {
        Assert.Equal("foo", _asyncLocal1.Value);
    }

    [Fact]
    public void WorksWithAsyncSetup() {
        Assert.Equal("foo", _asyncLocal2.Value);
    }

    public Task DisposeAsync() {
        return Task.CompletedTask;
    }
}
```

`WorksWithSyncSetup` passes consistently; `WorksWithAsyncSetup` fails consistently. I'm hoping this example eliminates some of the noise associated with `Task.Run`, parallel tests, inconsistent results, etc, and gets to the heart of the issue. Is this expected behavior? Thanks!

**Comment by bradwilson** [comment]
@tmenier Sounds like a legit bug, can you please open a new issue for this? Thanks!

**Comment by tmenier** [comment]
[Done](https://github.com/xunit/xunit/issues/1885), thanks

**Comment by Mauzer93** [comment]
@bradwilson  It's indeed correct. that AsyncLocal should flow to asynchronous calls, but changes made in those asynchronous calls never flow back to the caller. This is expected behavior.
However, in the code I experienced the issue, the AsyncLocal didn't flow the to asynchronous method somehow. I was not able to reproduce this bug and could not share the original implementation, as the code falls under a non disclosure agreement. I was able to fix it by using another implementation, however, I was never able to explain the actual bug.
